plugins/discovery-ec2/src/main/java/org/opensearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -32,82 +32,115 @@
+            .socketTimeout(Duration.ofMillis(readTimeoutMillis));
+
+        Ec2ClientBuilder builder = Ec2Client.builder()
+            .overrideConfiguration(overrideConfiguration)
+            .httpClientBuilder(clientBuilder)
+            .credentialsProvider(awsCredentialsProvider)
+            .region(region);
+
         if (Strings.hasText(endpoint)) {
             logger.debug("using explicit ec2 endpoint [{}]", endpoint);
-            builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, null));
+            builder.endpointOverride(URI.create(endpoint));
         }
+
         return SocketAccess.doPrivileged(builder::build);
     }
 
-    // pkg private for tests
-    static ClientConfiguration buildConfiguration(Logger logger, Ec2ClientSettings clientSettings) {
-        final ClientConfiguration clientConfiguration = new ClientConfiguration();
-        // the response metadata cache is only there for diagnostics purposes,
-        // but can force objects from every response to the old generation.
-        clientConfiguration.setResponseMetadataCacheSize(0);
-        clientConfiguration.setProtocol(clientSettings.protocol);
+    static ProxyConfiguration buildProxyConfiguration(Logger logger, Ec2ClientSettings clientSettings) {
         if (Strings.hasText(clientSettings.proxyHost)) {
             // TODO: remove this leniency, these settings should exist together and be validated
-            clientConfiguration.setProxyHost(clientSettings.proxyHost);
-            clientConfiguration.setProxyPort(clientSettings.proxyPort);
-            clientConfiguration.setProxyUsername(clientSettings.proxyUsername);
-            clientConfiguration.setProxyPassword(clientSettings.proxyPassword);
+            return ProxyConfiguration.builder()
+                .endpoint(URI.create(clientSettings.protocol + "://" + clientSettings.proxyHost + ":" + clientSettings.proxyPort))
+                .username(clientSettings.proxyUsername)
+                .password(clientSettings.proxyPassword)
+                .build();
+        } else {
+            return ProxyConfiguration.builder().build();
         }
-        // Increase the number of retries in case of 5xx API responses
-        clientConfiguration.setMaxErrorRetry(10);
-        clientConfiguration.setSocketTimeout(clientSettings.readTimeoutMillis);
-        return clientConfiguration;
+    }
+
+    static ClientOverrideConfiguration buildOverrideConfiguration(Logger logger, Ec2ClientSettings clientSettings) {
+        return ClientOverrideConfiguration.builder().retryPolicy(buildRetryPolicy(logger, clientSettings)).build();
     }
 
     // pkg private for tests
-    static AWSCredentialsProvider buildCredentials(Logger logger, Ec2ClientSettings clientSettings) {
-        final AWSCredentials credentials = clientSettings.credentials;
+    static RetryPolicy buildRetryPolicy(Logger logger, Ec2ClientSettings clientSettings) {
```

Comment:
**NOTE**: AWS SDK v2 introduced a concept of [TokenBucketRetryCondition](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/retry/conditions/TokenBucketRetryCondition.html), which effectively limits retries on a bucket level for APIs that have been failing continuously. It allocates tokens (default is 500), which means that once 500 retries fail for any API on a bucket, new retries will only be allowed once some retries are rejected. This is a deviation from AWS SDK v1. We might need some tuning here.

Ref: https://github.com/opensearch-project/OpenSearch/pull/7372/files#diff-b8fbfee9d71f7f2caf16514727692e968b4638afcafce12b4698e2046e6cb673R318

Reply:
Let's call it out in the PR. Adding a comment here.